src/aiida_quantumespresso/calculations/matdyn.py

```diff
@@ -47,18 +49,50 @@ def define(cls, spec):
             message='Number of kpoints in the inputs is not commensurate with those in the output')
         # yapf: enable
 
-    def _get_following_text(self):
-        """Add the kpoints after the namelist."""
-        try:
-            kpoints = self.inputs.kpoints.get_kpoints()
-        except AttributeError:
-            kpoints = self.inputs.kpoints.get_kpoints_mesh(print_list=True)
+    @staticmethod
+    def _validate_inputs(value, _):
+        """Validate the top level namespace."""
+        if 'parameters' in value:
+            parameters = value['parameters'].get_dict()
+        else:
+            parameters = {}
+
+        if parameters.get('INPUT', {}).get('flfrc', None) is not None:
+            return '`INPUT.flfrc` is set automatically from the `force_constants` input.'
+
+    def generate_input_file(self, parameters):
+        """Generate namelist input_file content given a dict of parameters.
+
+        :param parameters: 'dict' containing the fortran namelists and parameters to be used.
+          e.g.: {'CONTROL':{'calculation':'scf'}, 'SYSTEM':{'ecutwfc':30}}
+        :return: 'str' containing the input_file content a plain text.
+        """
+        kpoints = self.inputs.kpoints
+        append_string = ''
+
+        parameters.setdefault('INPUT', {})['flfrc'] = self.inputs.force_constants.filename
 
-        kpoints_string = [f'{len(kpoints)}']
-        for kpoint in kpoints:
-            kpoints_string.append('{:18.10f} {:18.10f} {:18.10f}'.format(*kpoint))  # pylint: disable=consider-using-f-string
+        # Calculating DOS requires (nk1,nk2,nk3), see
+        # https://gitlab.com/QEF/q-e/-/blob/develop/PHonon/PH/matdyn.f90#L72-73
+        if parameters['INPUT'].get('dos', False):
+            kpoints_mesh = kpoints.get_kpoints_mesh()[0]
```

Comment:
Should this have the same `try/except` as the `else` clause? Maybe take that try/except out of the if-block to do it only once

Reply:
Also, this is a non-related change, isn't it?

Reply:
I hate disentangling commit changes. 😭

Reply:
Removed the changes that should be in the other PR, now the try/except is gone